scripts/debug_tools/reduce.py

```diff
@@ -0,0 +1,304 @@
+
+    # --- Always do a clean analysis --- #
+
+    analyzer_cmd.append('--clean')
+
+    # --- Disable all checkers but the required one if needed --- #
+
+    if specific_checker is not None:
+        while remove_flag(analyzer_cmd, '-e'):
+            pass
+        while remove_flag(analyzer_cmd, '--enable'):
+            pass
+        while remove_flag(analyzer_cmd, '-d'):
+            pass
+        while remove_flag(analyzer_cmd, '--disable'):
+            pass
+
+        for category in checker_categories():
+            analyzer_cmd.extend(['-d', category])
+
+        analyzer_cmd.extend(['-e', specific_checker])
+
+    return analyzer_cmd
+
+
+def get_build_command_for_file(compilation_database, filename):
+    '''
+    This function finds the first build action in the compilation database
+    which compiles the given source file.
+    '''
+    with open(compilation_database) as f:
+        build_actions = json.load(f)
+
+    collection = map(
+        operator.itemgetter('command'),
+        filter(lambda ba: ba['file'].endswith(filename), build_actions))
+
+    return None if len(collection) == 0 else collection[0]
+
+
+def dump_preprocessed_build_action(build_command, output):
+    '''
+    This function dumps the preprocessed source code to the given output file.
+    '''
+
+    if not isinstance(build_command, list):
+        build_command = build_command.split()
+
+    build_command.append('-E')
+
+    try:
+        build_command[build_command.index('-o') + 1] = output
+    except ValueError:
+        build_command.extend(['-o', output])
+
+    print(build_command)
+    subprocess.Popen(build_command).communicate()
+
+
+def write_creduce_test_file(analyze_cmd, report_dir, parse_content, output_file):
+    '''
+    This function assembles and writes the test file which describes the
+    condition of recuding the source file with CReduce.
+    '''
+    content = '''
+#!/usr/bin/env bash
+{0}
+sed -i 's/, "working_directory": "[^"]*"//' {1}/metadata.json
+CodeChecker parse --print-steps {1} | grep '{2}'
+'''.format(analyze_cmd, report_dir, parse_content)
+
+    with open(output_file, 'w') as f:
+        f.write(content)
+
+    os.chmod(output_file, stat.S_IXUSR | stat.S_IRUSR)
+
+
+def get_prepared_build_action(workdir, build_action, reduce_file):
+    source_file = os.path.join(workdir, reduce_file)
+    return {
+        'directory': build_action['directory'],
+        'command': re.sub('[\w\/\.]*' + reduce_file,
+                          source_file,
+                          build_action['command']),
+        'file': source_file}
+
+
+# --- Command line arguments --- #
+
+parser = argparse.ArgumentParser(description='''
+Reduce for a specific bug.
+
+Given the following arguments this script attempts to create a minimal example
+which still produces a specific CodeChecker finding. Reduction steps are
+accomplished as long as the bug is still observable based on the output of
+CodeChecker parse command.''')
```

Comment:
Could you please write some usage examples in the epilog:
```py
parser = argparse.ArgumentParser(description='''...''',
                                 formatter_class=argparse.RawDescriptionHelpFormatter,
                                 epilog='''
Example scenario: Reduce source code
------------------------------------------------
TODO: write some usage example
''')
```
For example we use it at cmd:
https://github.com/Ericsson/codechecker/blob/7a28027bf19ad2348eba25455522f3021cd017ff/libcodechecker/libhandlers/cmd.py#L933